lib/view_component/base.rb

```diff
@@ -284,6 +349,47 @@ def with_content_areas(*areas)
         self.content_areas = areas
       end
 
+      # support initalizing slots as:
+      #
+      # with_slots :header
+      def with_slots(*slots)
+        # Generate a new slot class based on symbol
+        # For example: :header => MyComponent::Header < ViewComponent::Slot
+        slots.each do |slot|
+          self.class_eval("class #{slot.to_s.capitalize} < ViewComponent::Slot; end")
+        end
+      end
+
+      # support initalizing collection slots as:
+      #
+      # with_collection_slots :tab
+      def with_collection_slots(*slots)
+        # Generate a new slot class based on symbol
+        # For example: :tab => MyComponent::Tab < ViewComponent::CollectionSlot
+        slots.each do |slot|
+          self.class_eval("class #{slot.to_s.capitalize} < ViewComponent::CollectionSlot; end")
+        end
+      end
+
+      # register a Slot
+      #
+      # Called by `Slotable.inherited`
+      def register_slot(klass)
+        if klass.name.demodulize.to_s == "Content"
+          raise ArgumentError.new ":content is a reserved slot name. Please use another name, such as ':body'"
+        end
+
+        if klass.superclass == ViewComponent::CollectionSlot
+          # If slot is a collection, pluralize the name of the slot (Item => items)
+          attr_accessor ActiveSupport::Inflector.pluralize(klass.name.demodulize.downcase).to_sym
+        else
+          attr_accessor klass.name.demodulize.downcase.to_sym
+        end
```

Comment:
It feels a little strange to have effectively a switch statement here vs it being OO. Couldn't the Slot classes declare this? That way consumers of the gem could create their own slot classes and override things like this.